### PR TITLE
feat: add agent log streaming and follow provisioner format

### LINKED ISSUE #7767: Enable startup log streaming in `coder ssh`
When `startup_script_behavior => blocking` (https://github.com/coder/coder/issues/7757) or workspace agent lifecycle is `startup_error`, we want to stream/show the startup log the same way provisioning logs are currently streamed.

- [ ] Update messaging to match provisioner log output
- [ ] Show startup script log when `startup_script_behavior => blocking`
- [ ] Show startup script log when lifecycle is `start_error`

Example (blocking):

```
✔ Setting up [2805ms]
> ⧗  Planning infrastructure
  data.coder_provisioner.me: Refreshing...
  ...
✔ Starting workspace [3726ms]
✔ Cleaning Up [26ms]
✔ Waiting for workspace agent to connect [500ms]
> ⧗  Running workspace agent startup script
  + echo Running apt-get update
  Running apt-get update
  + sudo apt-get update
  ...
  + echo Installing iperf3
  Installing iperf3
  + sudo apt-get install iperf3 --yes
  Reading package lists... Done
  ...
✔ Running workspace agent startup script [999999ms]

user@workspace #
```

Example (non-blocking), when logging in before workspace agent has connected (this allows us to give users a hint whilst not changing any of the current functionality):

```
...
✔ Starting workspace [3726ms]
✔ Cleaning Up [26ms]
✔ Waiting for workspace agent to connect [500ms]
> ⧗  Running workspace agent startup script (non-blocking)
  Notice: Startup script is still running and some things may be missing or not fully set up
  For more information and troubleshooting, see https://coder.com/docs/coder-oss/latest/templates#startup-sccript-is-still-executing

user@workspace #
```

Example (error):

```
✘ Running workspace agent startup script [999999ms]
  + echo Running apt-get update
  Running apt-get update
  + sudo apt-get update
  ...
  + echo Installing iperf3
  Installing iperf2
  + sudo apt-get install iperf2 --yes
  ...
  E: Unable to locate package iperf2

  Warning: The startup script exited with an error and your workspace may be in an incomplete state, you may inspect the log output above for what went wrong
  For more information and troubleshooting, see https://coder.com/docs/v2/latest/templates#agent-does-not-become-ready

user@workspace #
```

### PR DESCRIPTION
**Pull request by mafredri**
This PR unifies the output between provisioner jobs and waiting on the agent via a simple shared struct, based on the concept of "stages" from provisioner jobs.

The following changes have been implemented:

- Refactor "stage logger" from provisioner jobs into re-usable component
- Print `coder ssh` connecting status in the same way as provisioner jobs
- Show startup log in blocking and non-blocking mode (partial)
- Show timestamps for logs (both provisioner and ssh)

The following has been removed:

- Support for showing "don't panic" after a timeout
- Support for messaging about "startup script timeout", we can re-introduce it if it becomes a requested feature, but for now it adds complexity for very little gain in this new format

The tests are a good place to inspect expected output sequences in different scenarios.

Fixes #7767

image

Here's an example output stream from logging in during provision:

image

### CODE REVIEW

File: cli/cliui/provisionerjob.go [hunk 1]
```diff
@@ -194,34 +179,89 @@ func ProvisionerJob(ctx context.Context, writer io.Writer, opts ProvisionerJobOp
...
+	s.end(stage, duration, true)
+}
+
+func (s *stageWriter) Fail(stage string, duration time.Duration) {
+	s.flushLogs()
+	s.end(stage, duration, false)
+}
+
+//nolint:revive
+func (s *stageWriter) end(stage string, duration time.Duration, ok bool) {
```
Comment (mtojek) [inline cli/cliui/provisionerjob.go:220]: I like the way you refactored the code into `stageWriter`. It's really clean approach 👍

File: cli/cliui/agent.go [hunk 2]
```diff
@@ -2,31 +2,21 @@ package cliui
...
 	"golang.org/x/xerrors"
 
 	"github.com/coder/coder/codersdk"
 )
 
-var (
-	AgentStartError   = xerrors.New("agent startup exited with non-zero exit status")
-	AgentShuttingDown = xerrors.New("agent is shutting down")
-)
+var AgentShuttingDown = xerrors.New("agent is shutting down")
```
Comment (mtojek) [inline cli/cliui/agent.go:14]: nit: does it have to be exposed?
nit2: `ErrAgentShuttingDown` ?

File: cli/cliui/agent.go [hunk 3]
```diff
@@ -35,230 +25,217 @@ func Agent(ctx context.Context, writer io.Writer, opts AgentOptions) error {
 	if opts.FetchInterval == 0 {
 		opts.FetchInterval = 500 * time.Millisecond
 	}
-	if opts.WarnInterval == 0 {
-		opts.WarnInterval = 30 * time.Second
+	if opts.Wait && opts.FetchLogs == nil {
+		return xerrors.Errorf("fetch logs required when waiting for agent")
```
Comment (mtojek) [inline cli/cliui/agent.go:29]: nit: I'm not sure if I understand what this error means or whether it is actionable. Users may have trouble understanding it too.

**Review by mtojek — APPROVED** [review]
I haven't found anything blocking here, but it was much easier to review by reading the target file [agent.go](https://github.com/coder/coder/blob/d42bf8c0579c5897fc07629f6866758a2f7ce382/cli/cliui/agent.go) :)

File: cli/cliui/agent.go (hunk 2 above)
Comment (mafredri) [reply cli/cliui/agent.go:14]: Ah, you're right. Seems we're not testing for the error currently so we can unexport it. 👍🏻

**Comment by mtojek** [comment]
Let's not wait for other reviews. This is already in good shape.

**Pull request merged**